src/algebra/order/to_interval_mod.lean

```diff
@@ -457,6 +458,106 @@ begin
     rw [hz, to_Ioc_mod_zsmul_add] }
 end
 
+
+/-! ### Links between the `Ico` and `Ioc` variants applied to the same elemnt -/
+section Ico_Ioc
+
+/-- `to_Ico_div` agrees with `to_Ioc_div` if `x` does not lie on a boundary. -/
+lemma to_Ico_div_eq_to_Ioc_div (a : α) {b : α} (hb : 0 < b) (x : α) :
+  to_Ico_div a hb x = to_Ioc_div a hb x ↔ ∀ z : ℤ, x ≠ a + z • b :=
```

Comment:
I needed the LHS [here](https://github.com/leanprover-community/mathlib/compare/identify_interval_ends...jyxu/identify_interval_ends?expand=1#diff-289dc9987f8c37ae4d6fc491bf58a0d8446444f6cedc0e819d27e46379dc8ae9R428), but under the condition `to_Ioc_mod a hp.out x ≠ a + p` instead. Maybe my proof could inspire some golfs ...

Reply:
That link doesn't seem to link to a specific line to me. I'm not sure what the best way to state this lemma is, there are two many equivalent versions

Reply:
Yes the link doesn't make the line(s) highlighted, but jumps to the correct place. With the Refined GitHub enabled, however, it behaves a bit erratically (jumps to the correct place + 2 lines). My proof of `to_Ico = to_Ioc` there is just two lines (428-429) in `Icc_circle_equiv_comp_mk'`.

Reply:
> And we should probably match the statement of to_Ico_mod_eq_iff and use sub instead of add

Note these are about `x` and `y` not `x` and `a`, so the matching you're thinking of isn't particularly meaningful.

Reply:
I suggest that you prove `tfae [to_Ico_div a hb x = to_Ioc_div a hb x, to_Ico_mod a hb x = to_Ioc_mod a hb x, to_Ioc_mod a hb x ≠ a + b, to_Ico_mod a hb x ≠ a, ∀ z : ℤ, a - x ≠ z • b]`. The last two are equivalent by [to_Ico_mod_eq_iff](https://leanprover-community.github.io/mathlib_docs/algebra/order/to_interval_mod.html#to_Ico_mod_eq_iff) and push_neg.

Reply:
Do we have 
```
instance ordered_smul.no_smul_divisors {R M} [linear_ordered_semiring R]
  [linear_ordered_add_comm_monoid M] [smul_with_zero R M] [ordered_smul R M] :
  no_zero_smul_divisors R M :=
```
? I'm trying to use [smul_left_injective](https://leanprover-community.github.io/mathlib_docs/algebra/module/basic.html#smul_left_injective) (there's already [int.ordered_smul](https://leanprover-community.github.io/mathlib_docs/algebra/order/smul.html#int.ordered_smul)) but I guess it can be replaced with `zsmul_strict_mono_left`.

Reply:
Hmm, [add_right_inj](https://leanprover-community.github.io/mathlib_docs/algebra/group/defs.html#add_right_inj) is the same as add_left_cancel_iff except for implicitness.

Reply:
Now here's a huge TFAE that encompasses most of the following lemmas:
```
lemma tfae_to_Ico_eq_to_Ioc (a : α) {b : α} (hb : 0 < b) (x : α) :
  [to_Ico_div a hb x = to_Ioc_div a hb x,
   to_Ico_div a hb x ≠ to_Ioc_div a hb x - 1,
   to_Ico_mod a hb x = to_Ioc_mod a hb x,
   to_Ico_mod a hb x ≠ to_Ioc_mod a hb x - b,
   to_Ioc_mod a hb x ≠ a + b,
   to_Ico_mod a hb x ≠ a,
   ∀ z : ℤ, a - x ≠ z • b].tfae :=
begin
  tfae_have : 1 ↔ 3,
  { rw [to_Ico_mod, to_Ioc_mod, add_right_inj, (zsmul_strict_mono_left hb).injective.eq_iff] },
  tfae_have : 1 → 2,
  { intro h, linarith },
  tfae_have : 2 ↔ 4,
  { rw [ne, ne, to_Ico_mod, to_Ioc_mod, ← add_sub, add_right_inj],
    convert iff.not (zsmul_strict_mono_left hb).injective.eq_iff.symm,
    rw [sub_smul, one_smul] },
  tfae_have : 4 → 5,
  { refine mt (λ h, _), rw h,
    rw [to_Ico_mod_eq_iff, add_sub_cancel],
    refine ⟨le_rfl, lt_add_of_pos_right a hb, to_Ioc_div a hb x - 1, _⟩,
    conv_lhs { rw [← sub_eq_iff_eq_add.2 h, to_Ioc_mod] },
    rw [sub_smul, one_smul], abel },
  tfae_have : 5 → 3,
  { have := to_Ioc_mod_mem_Ioc a hb x,
    exact λ h, (to_Ico_mod_eq_iff hb).2 ⟨this.1.le, this.2.lt_of_ne h, _, add_sub_cancel' x _⟩ },
  tfae_have : 3 ↔ 6,
  { rw [eq_comm, to_Ioc_mod_eq_iff],
    have := to_Ico_mod_mem_Ico a hb x,
    refine ⟨λ h, h.1.ne', λ h, ⟨this.1.lt_of_ne' h, this.2.le, _, add_sub_cancel' x _⟩⟩ },
  tfae_have : 6 ↔ 7,
  { rw ← not_iff_not, push_neg, split; intro h,
    { rw ← h, exact ⟨_, add_sub_cancel' x _⟩ },
    { exact (to_Ico_mod_eq_iff hb).2 ⟨le_rfl, lt_add_of_pos_right a hb, h⟩ } },
  tfae_finish,
end
```

Reply:
`tfae_to_Ico_eq_to_Ioc` added to #17983

Reply:
I've rewritting my proofs to use these new lemmas where obvious. I haven't yet pruned the redundant lemmas.